Products/CMFPlone/relationhelper.py

```diff
@@ -132,71 +139,62 @@ def purge_relations(context=None):
 
 
 def restore_relations(context=None, all_relations=None):
-    """Restore relations from a annotation on the portal."""
+    """Restore relations from an annotation on the portal."""
 
     portal = getSite()
     if all_relations is None:
         all_relations = IAnnotations(portal)[RELATIONS_KEY]
-    logger.info(f"Loaded {len(all_relations)} relations to restore")
+    logger.info(f"Loaded {len(all_relations)} relations to restore.")
     update_linkintegrity = set()
     modified_items = set()
     modified_relation_lists = defaultdict(list)
 
-    # remove duplicates but keep original order
+    # Remove duplicates but keep original order.
     unique_relations = []
     seen = set()
-    seen_add = seen.add
     for rel in all_relations:
         hashable = tuple(rel.items())
         if hashable not in seen:
             unique_relations.append(rel)
-            seen_add(hashable)
+            seen.add(hashable)
         else:
             logger.info(f"Dropping duplicate: {hashable}")
 
     if len(unique_relations) < len(all_relations):
         logger.info(f"Dropping {len(all_relations) - len(unique_relations)} duplicates")
-        all_relations = unique_relations
 
+    # Update unique relations.
     intids = getUtility(IIntIds)
-    for index, item in enumerate(all_relations, start=1):
-        if not index % 500:
-            logger.info(f"Restored {index} of {len(all_relations)} relations...")
-
-        # source object for UID exists.
-        try:
+    new_index = 0
+    for index, item in enumerate(unique_relations, start=1):
+        # Get source object for UID. Skip relation if no object found.
+        if item["from_uuid"] is None:
+            logger.warning(f"No source object. {tuple(item.items())}.")
+            continue
+        else:
             source_obj = uuidToObject(item["from_uuid"])
-        except KeyError:
-            # brain exists but no object
-            source_obj = None
-        finally:
-            if not source_obj:
-                logger.info(f'No source object found for UID {item["from_uuid"]}.')
-                continue
```

Comment:
Why are you removing the try/except KeyError here and further on?  If the portal catalog or relation catalog are out of sync with the object database or with each other, you may very well see KeyErrors.

I suppose if you call `rebuild_relations` it will work, because that calls `store_relations` and `restore_relations` after one another.  But if in custom code you call `store_relations`, then remove a content item with a relation, and then call `restore_relations`, you will get a KeyError.

Reply:
@ksuess Does this case still need to be handled?

Reply:
> @ksuess Does this case still need to be handled?

@david I'll continue working in this today, writing tests and responding to change requests.

Reply:
@mauritsvanrees I could be wrong, but I do not see how a KeyError should be raised in https://github.com/plone/Products.CMFPlone/pull/3758/files/d3c50e2a60091b933aa8517c692c1233407ba8dd..dabed3cf67bf7e27bb361d62fbfca6b6a6ff9d59#diff-6d53467a1d2dd42d829f6bf8c87ed4a32aa04dea987f2d92ed34ed4a33c87f28L168 
If there is no object found, uuidToObject returns _None_.
I added a test. 
https://github.com/plone/Products.CMFPlone/pull/3758/files#diff-1bade6b3c76ae89fc7f34cf3acb70cefdfb738529c4ae263d69ad258750a8c6bR141-R178
Please have a look if I understood your objection correctly or tell me what you mean instead. Thank you.

Reply:
`uuidToObject` calls `uuidToPhysicalPath`.  This tries to get an "rid" ('reverse id' I think) from the UID index.  If this gives a wrong answer, then the [call to `catalog.getpath`](https://github.com/plone/plone.app.uuid/blob/2.2.2/plone/app/uuid/utils.py#L57) will give a KeyError:

```
>>> app.Plone.portal_catalog._catalog.getpath(12345)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/maurits/shared-eggs/cp311/Products.ZCatalog-7.0-py3.11.egg/Products/ZCatalog/ZCatalog.py", line 520, in getpath
    return self._catalog.paths[rid]
           ~~~~~~~~~~~~~~~~~~~^^^^^
KeyError: 12345
```

No, this should never happen.
But yes, I _have_ seen this happen in practice, and have written a script to [fix such a broken catalog](https://github.com/zestsoftware/plonescripts/blob/master/fix_uid_index.py). According to the readme in that repo: "This can at least happen when you import a zexp twice, in different folders."

So I would prefer to keep the `try/except KeyError`. :-)

Reply:
Done.
Catch KeyError of uuidToObject: uuid is indexed in UID, but catalog cannot resolve rid to path.
https://github.com/plone/Products.CMFPlone/pull/3758/commits/fe1ed81c9a7eb19b889ff319e51204ce040c3ff5